[config] Exit with non-zero when qos reload fail

jianyuewu opened the pull request:
Fix issue where 'config qos reload' returns exit code 0, when operation unsuccessful, now it return 1 after executing all the commands.

<!--
    Please make sure you've read and understood our contributing guidelines:
    https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

    ** Make sure all your commits include a signature generated with `git commit -s` **

    If this is a bug fix, make sure your description includes "closes #xxxx",
    "fixes #xxxx" or "resolves #xxxx" so that GitHub automatically closes the related
    issue when the PR is merged.

    If you are adding/modifying/removing any command or utility script, please also
    make sure to add/modify/remove any unit tests from the tests
    directory as appropriate.

    If you are modifying or removing an existing 'show', 'config' or 'sonic-clear'
    subcommand, or you are adding a new subcommand, please make sure you also
    update the Command Line Reference Guide (doc/Command-Reference.md) to reflect
    your changes.

    Please provide the following information:
-->

#### What I did
When config qos reload failed with error print "Operation not completed successfully", exit with code 1 finally, instead of exit with code 0.

#### How I did it
Add WAIT_UNTIL_CLEAR_STATUS flag in function _wait_until_clear() when error reported, and after execute command finish, check this flag, if it is none zero, then exit 1.

#### How to verify it
Test with command "config qos reload". Note that "config load_minigraph" is also verified, not impacted.

#### Previous command output (if the output of a command-line utility has changed)
```bash
# config qos reload
Operation not completed successfully, please save and reload configuration.
Running command: /usr/local/bin/sonic-cfggen -d -t /usr/share/sonic/device/x86_64-nvidia_sn5600-r0/Mellanox-SN5600-C256S1/buffers_dynamic.json.j2,/tmp/cfg_buffer.json -t /usr/share/sonic/device/x86_64-nvidia_sn5600-r0/Mellanox-SN5600-C256S1/qos.json.j2,/tmp/cfg_qos.json -y /etc/sonic/sonic_version.yml
Running command: /usr/local/bin/sonic-cfggen -j /tmp/cfg_buffer.json -j /tmp/cfg_qos.json --write-to-db
Buffer calculation model updated, restarting swss is required to take effect
```
#### New command output (if the output of a command-line utility has changed)
```bash
# config qos reload
Operation not completed successfully, please save and reload configuration.
Running command: /usr/local/bin/sonic-cfggen -d -t /usr/share/sonic/device/x86_64-nvidia_sn5600-r0/Mellanox-SN5600-C256S1/buffers_dynamic.json.j2,/tmp/cfg_buffer.json -t /usr/share/sonic/device/x86_64-nvidia_sn5600-r0/Mellanox-SN5600-C256S1/qos.json.j2,/tmp/cfg_qos.json -y /etc/sonic/sonic_version.yml
Running command: /usr/local/bin/sonic-cfggen -j /tmp/cfg_buffer.json -j /tmp/cfg_qos.json --write-to-db
Buffer calculation model updated, restarting swss is required to take effect
# echo $?
1
```

keboliu commented on config/main.py:
```diff
@@ -3227,18 +3232,23 @@ def reload(ctx, no_dynamic_buffer, no_delay, dry_run, json_data, ports, verbose)
-                clicommon.run_command(command, display_cmd=True)
+                out, rc = clicommon.run_command(command, display_cmd=True, return_cmd=True)
+                if rc != 0:
+                    click.echo("Command failed due to rc value!")
```
this output is not informative to the user

jianyuewu replied:
Agree, will update it with: click.echo(f"Config reload failed: {' '.join(command)} with return code: {rc}, please retry")

jianyuewu replied:
Looks this extra log print is not needed, as it will automatically return none 0 value, when it fails.

keboliu commented on config/main.py:
```diff
@@ -3252,6 +3262,10 @@ def reload(ctx, no_dynamic_buffer, no_delay, dry_run, json_data, ports, verbose)
         print("Buffer calculation model updated, restarting swss is required to take effect")
 
+    if WAIT_UNTIL_CLEAR_STATUS != 0:
+        click.echo("Command failed due to _wait_until_clear() failed!")
```
Is this inner function name  _wait_until_clear()  helpful to the user to understand the failure? suggest to revise

jianyuewu replied:
Agree, updated in new patch, with click.echo("Some keys in APPL_DB didn't update successfully, please reload configuration.")

keboliu commented on config/main.py:
```diff
@@ -3227,18 +3232,23 @@ def reload(ctx, no_dynamic_buffer, no_delay, dry_run, json_data, ports, verbose)
-                    clicommon.run_command(command + ["--write-to-db"], display_cmd=True)
+                    out, rc = clicommon.run_command(command + ["--write-to-db"], display_cmd=True, return_cmd=True)
+                if rc != 0:
+                    click.echo("Command failed due to rc value!")
```
same comment as above one

jianyuewu replied:
Yes, this extra log print is not needed, as it will automatically return none 0 value, when it fails, so still keep as original and only modify when _wait_until_clear() update APPL_DB keys failed.

stephenxs commented on config/main.py:
```diff
@@ -795,6 +797,7 @@ def _wait_until_clear(tables, interval=0.5, timeout=30, verbose=False):
     if not empty:
         click.echo("Operation not completed successfully, please save and reload configuration.")
+        WAIT_UNTIL_CLEAR_STATUS = 1
     return empty
```
is it possible to make use of this return value instead of a global variable?

jianyuewu replied:
Agree, good idea, will try it.

jianyuewu committed:
[config] Exit with non-zero when qos reload fail

Fix issue when 'config qos reload' operation unsuccessful,
but returns exit code 0, when operation unsuccessful,
now it will return code 1 after executing all the commands.

Signed-off-by: Jianyue Wu <jianyuew@nvidia.com>

jianyuewu commented:
Code updated, could you help review again? Thank you~ UT is still ongoing.

stephenxs commented on config/main.py:
```diff
@@ -795,6 +795,7 @@ def _wait_until_clear(tables, interval=0.5, timeout=30, verbose=False):
         empty = (non_empty_table_count == 0)
     if not empty:
         click.echo("Operation not completed successfully, please save and reload configuration.")
+        return False
```
looks like unnecessary... can fallback?

jianyuewu replied:
Agree, looks unnecessary, will remove it.

stephenxs commented on config/main.py:
```diff
@@ -3169,7 +3170,7 @@ def reload(ctx, no_dynamic_buffer, no_delay, dry_run, json_data, ports, verbose)
     log.log_info("'qos reload' executing...")
     if not dry_run:
-        _clear_qos(delay = not no_delay, verbose=verbose)
+        clear_qos_status = _clear_qos(delay = not no_delay, verbose=verbose)
```
maybe just `status` suffice.
in the future perhaps we will have other steps that can fail.

jianyuewu replied:
Agree, will use status directly.

stephenxs commented on config/main.py:
```diff
@@ -3252,6 +3253,10 @@ def reload(ctx, no_dynamic_buffer, no_delay, dry_run, json_data, ports, verbose)
     if buffer_model_updated:
         print("Buffer calculation model updated, restarting swss is required to take effect")
 
+    if clear_qos_status != True:
```
```suggestion
    if not clear_qos_status:
```

jianyuewu replied:
OK, will use if not status instead.

jianyuewu commented:
Code updated, could you help review? Thank you~😊

stephenxs commented on config/main.py:
```diff
@@ -3252,6 +3252,10 @@ def reload(ctx, no_dynamic_buffer, no_delay, dry_run, json_data, ports, verbose)
         print("Buffer calculation model updated, restarting swss is required to take effect")
 
+    if not status:
+        click.echo("Some keys in APPL_DB didn't update successfully, please reload configuration.")
```
I'm not sure whether we still need message here since we already printed here https://github.com/jianyuewu/sonic-utilities/pull/4/files#diff-8bb4eed8e9e1f9974cbb7b46799411ecc318bae391f48735e7ad84a8122809a0R797
@keboliu  what do you think?

jianyuewu replied:
Agree, there is print previously, should be already informative, and latter has sys.exit(1), should be not needed.
